### PR TITLE
[Feature] Support edge feat reconstruct in multi-task learning.

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*

*Description of changes:*
Add support for edge feat reconstruct in multitask learning. We treat edge feature reconstruction as an regression task.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/gsf.py [hunk 1]
```diff
@@ -338,6 +340,46 @@ def create_builtin_reconstruct_nfeat_decoder(g, decoder_input_dim, config, train
...
+        according to task config
+
+    Parameters
+    ----------
+    g: DGLGraph
+        The graph data.
+        Note(xiang): Make it consistent with create_builtin_edge_decoder.
+        Reserved for future.
+    decoder_input_dim: int
+        Input dimension size of the decoder
```
Comment (zhjwy9343) [inline python/graphstorm/gsf.py:355]:
```suggestion
        Input dimension size of the decoder.
```

File: python/graphstorm/gsf.py [hunk 2]
```diff
@@ -338,6 +340,46 @@ def create_builtin_reconstruct_nfeat_decoder(g, decoder_input_dim, config, train
...
+    config: GSConfig
+        Configurations
```
Comment (zhjwy9343) [inline python/graphstorm/gsf.py:357]:
```suggestion
        Configurations.
```

File: python/graphstorm/gsf.py [hunk 3]
```diff
@@ -338,6 +340,46 @@ def create_builtin_reconstruct_nfeat_decoder(g, decoder_input_dim, config, train
...
+    train_task : bool
+        Whether this model is used for training.
+
+    Returns
+    -------
+    decoder: The node task decoder(s)
```
Comment (zhjwy9343) [inline python/graphstorm/gsf.py:363]:
```suggestion
    decoder: The edge task decoder(s)
```

File: python/graphstorm/gsf.py [hunk 4]
```diff
@@ -338,6 +340,46 @@ def create_builtin_reconstruct_nfeat_decoder(g, decoder_input_dim, config, train
...
+    loss_func: The loss function(s)
+    """
+    dropout = config.dropout if train_task else 0
+    # Only support on edge type per reconstruction type
```
Comment (zhjwy9343) [inline python/graphstorm/gsf.py:367]:
```suggestion
    # Only support one edge type for reconstruction tasks
```

File: python/graphstorm/gsf.py [hunk 5]
```diff
@@ -338,6 +340,46 @@ def create_builtin_reconstruct_nfeat_decoder(g, decoder_input_dim, config, train
...
+    target_etype = config.target_etype[0]
+    reconstruct_feat = config.reconstruct_efeat_name
+    assert len(g.edges[target_etype].data[reconstruct_feat].shape) == 2, \
+        "The edge feature {reconstruct_feat} of {target_etype} edges " \
+        f"Must be a 2D tensor, but got {g.edges[target_etype].data[reconstruct_feat].shape}"
```
Comment (zhjwy9343) [inline python/graphstorm/gsf.py:372]:
```suggestion
        f"Must be a 2D tensor, but got {g.edges[target_etype].data[reconstruct_feat].shape}."
```

File: python/graphstorm/gsf.py [hunk 6]
```diff
@@ -338,6 +340,46 @@ def create_builtin_reconstruct_nfeat_decoder(g, decoder_input_dim, config, train
...
+# pylint: disable=unused-argument
+def create_builtin_reconstruct_efeat_decoder(g, decoder_input_dim, config, train_task):
+    """ create builtin edge feature reconstruction decoder
+        according to task config
+
+    Parameters
+    ----------
+    g: DGLGraph
+        The graph data.
+        Note(xiang): Make it consistent with create_builtin_edge_decoder.
```
Comment (zhjwy9343) [inline python/graphstorm/gsf.py:352]: In the #370 line, the ``g`` is used. So why here say, it is reserved for future?

**Review by zhjwy9343 — COMMENTED** [review]
1st batch of review comments.

File: python/graphstorm/config/argument.py [hunk 7]
```diff
@@ -620,6 +657,14 @@ def verify_node_feat_reconstruct_arguments(self):
         _ = self.eval_metric
         _ = self.reconstruct_nfeat_name
 
+    def verify_edge_feat_reconstruct_arguments(self):
+        """Verify the correctness of arguments for edge feature reconstruction tasks.
+        """
+        _ = self.target_ntype
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/config/argument.py:663]: Yes

File: tests/unit-tests/test_config.py [hunk 8]
```diff
@@ -2058,6 +2081,36 @@ def test_multi_task_config():
         assert nfr_config.eval_metric[0] == "rmse"
         assert nfr_config.batch_size == 64
 
+        # reconstruct edge feat
+        efr_config = config.multi_tasks[8]
+        assert efr_config.task_type == BUILTIN_TASK_RECONSTRUCT_EDGE_FEAT
+        print(efr_config.task_id)
```
Comment (zhjwy9343) [inline tests/unit-tests/test_config.py:2087]: Should we remove this print clause?

File: python/graphstorm/model/edge_decoder.py [hunk 9]
```diff
@@ -305,6 +305,167 @@ def out_dims(self):
...
+        h: dict of Tensor
+            The input node embeddings in the format of {ntype: emb}.
+        e_h: dict of Tensor
+            The input edge embeddings in the format of {(src_ntype, etype, dst_ntype): emb}.
+            Not used, but reserved for future support of edge embeddings. Default: None.
+
+        Returns
+        -------
+        out: Tensor
+            The prediction results. If this decoder is for edge classification, return the
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/model/edge_decoder.py:459]: Updated.

File: python/graphstorm/config/argument.py [hunk 10]
```diff
@@ -2911,9 +2958,18 @@ def reconstruct_nfeat_name(self):
         """ node feature name for reconstruction
         """
         assert hasattr(self, "_reconstruct_nfeat_name"), \
-            "reconstruct_nfeat_name must be provided under reconstruct_node_feat task "
+            "reconstruct_nfeat_name must be provided for reconstruct_node_feat tasks(s)."
         return self._reconstruct_nfeat_name
 
+    ################## Reconstruct edge feats ###############
+    @property
+    def reconstruct_efeat_name(self):
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/config/argument.py:2968]: Updated impl.

**Review by zhjwy9343 — APPROVED** [review]
LGTM

**Pull request merged**